index.js

```diff
@@ -124,19 +124,31 @@ function Encoder() {}
  */
 
 Encoder.prototype.encode = function(obj, callback){
-  if ((obj.type === exports.EVENT || obj.type === exports.ACK) && hasBin(obj.data)) {
-    obj.type = obj.type === exports.EVENT ? exports.BINARY_EVENT : exports.BINARY_ACK;
+
+  /**
+   * For object data with circular references, hasBin will throw
+   * exceptions, which must be handled or it could hang the channel.
+   * And the circular references only happen to non-Binary data, 
+   * so we will invoke callback with an array.
+   */
+  try {
+    var hasBinary = hasBin(obj.data);
+  } catch (e) {
+    callback([encodeError()]);
```

Comment:
I'm wondering whether we should send the `encodeError()` packet to the other side.

Reply:
I think circular reference is something that should be handled at compile time, because a large trunk of data could be discarded because of 1 misplaced object pointer. I can improve the error message to make it more specific on what's the cause of this error, so developers can trace the problem better.